BAP unicast group add/remove streams

Thalley opened the pull request:
Add support for adding and removing streams to a non-started unicast group. 

Also fixes some check in the `bt_audio_unicast_group_delete` function. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41196

Thalley committed:
Bluetooth: Audio: Add support for add/rem streams for unicast group

Add a function to add more and remove streams
to and from a unicast group.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Fix valid checks for bt_audio_unicast_group_delete

The checks for stream->group was useless, and the check for
an active CIG was missing.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on include/bluetooth/audio/audio.h:
```diff
@@ -1516,6 +1516,45 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_audio_unicast_group_add_streams(struct bt_audio_unicast_group *unicast_group,
+				       struct bt_audio_stream *streams,
```
The ISO APIs takes arrays of pointers to allow for a more flexible API, so that the channels do not have to be in an array. Should we update the Audio API to the same?

asbjornsabo replied:
I would need some discussion to have an opinion on that.

Thalley replied:
Some details for the discussion: 

The ISO API uses an array of pointers, such you can e.g. allocate your ISO channels as 
```c
struct bt_iso_chan left;
struct bt_iso_chan right;

cig_param.cis_channels[0] = &left;
cig_param.cis_channels[1] = &right;

int bt_iso_cig_create(&cig_param, &cig);
```

whereas this takes an array of channels, so that you need to do
```c
struct bt_audio_stream streams[2];
struct bt_audio_stream *left = &streams[0];
struct bt_audio_stream *right = &streams[1];

bt_audio_unicast_group_add_streams(group, streams);
```

In other words, the channels need to be allocated in an array, which is less flexible IMO, especially if you need to allocate multiple streams per remove device. The alternative here would be 

```c
struct bt_audio_stream *streams[2];
struct bt_audio_stream left;
struct bt_audio_stream right;

streams[0] = &left;
streams[1] = &right;

bt_audio_unicast_group_add_streams(group, streams, ARRAY_SIZE(streams));
```

Thalley replied:
> But that would be a separate PR, I assume.

Indeed, I'll create an issue for it

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42587

Casper-Bonde-Bose commented on include/bluetooth/audio/audio.h:
```diff
@@ -1516,6 +1516,45 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+ *  This function can be used to add additional streams to a
+ *  bt_audio_unicast_group.
+ *
+ *  This can be called at any time before when none of the streams in the
+ *  group has not been started (see bt_audio_stream_ops.started()).
```
Typo:
```suggestion
 *  This can be called at any time before any of the streams in the
 *  group has been started (see bt_audio_stream_ops.started()).
```

Casper-Bonde-Bose commented on include/bluetooth/audio/audio.h:
```diff
@@ -1516,6 +1516,45 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+ *
+ *  This function can be used to remove streams from a bt_audio_unicast_group.
+ *
+ *  This can be called at any time before when none of the streams in the
+ *  group has not been started (see bt_audio_stream_ops.started()).
```
```suggestion
 *  This can be called at any time before any of the streams in the
 *  group has been started (see bt_audio_stream_ops.started()).
```

Casper-Bonde-Bose commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1070,21 +1070,127 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+	 * that would start the ISO connection procedure
+	 */
+	cig = unicast_group->cig;
+	if (cig != NULL && cig->state != BT_ISO_CIG_STATE_CONFIGURED) {
```
What about state BT_ISO_CIG_STATE_IDLE?

Casper-Bonde-Bose commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1070,21 +1070,127 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+	 * that would start the ISO connection procedure
+	 */
+	cig = unicast_group->cig;
+	if (cig != NULL && cig->state != BT_ISO_CIG_STATE_CONFIGURED) {
```
BT_ISO_CIG_STATE_IDLE?

Thalley replied:
`BT_ISO_CIG_STATE_IDLE` is only used for CIGs that have been been created. Once a CIG is created, it goes into the CONFIGURED state.

Casper-Bonde-Bose commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1070,21 +1070,127 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+	 * that would start the ISO connection procedure
+	 */
+	cig = unicast_group->cig;
+	if (cig != NULL && cig->state != BT_ISO_CIG_STATE_CONFIGURED) {
```
Same

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/42485#discussion_r800608869

Casper-Bonde-Bose reviewed (dismissed):
LGTM - a type and a question that is most likely just me not having enough insight into the code:-)

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1516,6 +1516,45 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+ *
+ *  This can be called at any time before any of the streams in the
+ *  group has been started (see bt_audio_stream_ops.started()).
+ *  Once all streams in a group has been stopped
```
Suggest to rephrase to "Additional streams can also be added using this function once all streams in the group has been stopped [...]" 
or "Additional streams can also be added after all streams in the groups have been stopped [...]"

Similar for the remove function.

Thalley replied:
`have been stopped` implying that the callback needs to have been called before you can call this, which isn't true. I'm up for rephrasing it, but it should be more accurate than this suggestion.

asbjornsabo replied:
Now I see that @Casper-Bonde-Bose also had comments on this (I was reviewing commit by commit).

I fail to see the difference regarding callbacks between what I commented on and my proposed text.
The "has been stopped" is there in both.  (It should be changed to a "have", though - "streams" is plural.)

My intention was only to reorder the sentence, so that the addition of streams comes first, and the condition next, for consistency with the previous sentence.
- streams may be added before any streams in the group have been started
- streams may be also added after all streams in the group have been stopped

Thalley replied:
> I fail to see the difference regarding callbacks between what I commented on and my proposed text.

> "Additional streams can also be added after all streams in the groups have been stopped [...]"

Implies that the streams have been stopped before this can be called, but it can also be called before they are started. Streams do not init to the the "stopped" state. 

But I think I see what you mean now. Let me try to rephrase it.

Thalley replied:
Let me know if the current wording is better.

jhedberg commented on include/bluetooth/audio/audio.h:
```diff
@@ -1514,6 +1514,45 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+ */
+int bt_audio_unicast_group_remove_streams(struct bt_audio_unicast_group *unicast_group,
+					  struct bt_audio_stream *streams,
+					  uint8_t num_stream);
```
`uint8_t` seems a bit arbitrary. Why not eg `size_t` as that’s the return value of `ARRAY_SIZE()`?

Thalley replied:
Used `uint8_t` as the maximum number of streams is 0x1F (as per the Core spec). 

I am fine with using size_t, but I will also add the maximum number of streams to the documentation (regardless of uint8_t/size_t) as that is missing.

Thalley replied:
Actually, it is already implicitly described by CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT, so won't add additional documentation (unless requested).

jhedberg replied:
I think I’d still prefer size_t since in the API context this relates to the previous array parameter and not over the air representation. And I suppose the functions anyway check for the max value and return error if too large.

jhedberg replied:
Btw, do we have some convention of using `array[]` instead of `*array` to make it clear the parameter is an array and not a pointer to a single element?

Thalley replied:
> Btw, don’t we have some convention of using `array[]` instead of `*array` to make it clear the parameter is an array and not a pointer to a single element?

We do not. I actually don't see any BT APIs use `array[]`.

jhedberg replied:
Good point. I took a look at e.g. `bt_le_adv_start()` (one of the "old" APIs we have), and it uses pointer syntax. Still, I don't see any reasons not to make the distinction clearer by using `[]`, but that'd then need to be done consistently throughout the subsystem (i.e. independent of this PR).

jhedberg replied:
> The value is used for a HCI request that takes an uint8_t however, so we'll need to add a cast later on if we pursue the size_t route.

Why would you need an (explicit) cast? I don't think the compiler will complain of an implicit case from size_t to uint8_t, will it?

jhedberg replied:
Furthermore, variable types of members of packed structs (such as HCI command parameters) should in many ways be thought of as belonging to a separate domain because of different alignment and byte order (for multi-byte types) assumptions, meaning one always needs to pay some attention when encoding to/from them.

jhedberg replied:
> Do you mind if I create an issue to fix this in a separate PR? This will also affect the ISO API which also use uint8_t for array counts.

Can't the ISO API be updated in a separate PR?

jhedberg replied:
Btw, can you point me to the specific APIs in the ISO API using uint8_t for array sizes?

Thalley replied:
> Btw, can you point me to the specific APIs in the ISO API using uint8_t for array sizes?

```c
struct bt_iso_cig_param {
	/** @brief Array of pointers to CIS channels */
	struct bt_iso_chan **cis_channels;

	/** @brief Number channels in @p cis_channels
	 *
	 *  Maximum number of channels in a single group is
	 *  BT_ISO_MAX_GROUP_ISO_COUNT
	 */
	uint8_t num_cis;
```

Thalley replied:
> > Do you mind if I create an issue to fix this in a separate PR? This will also affect the ISO API which also use uint8_t for array counts.
> 
> Can't the ISO API be updated in a separate PR?

Sure, we can also just do it that way :)

jhedberg replied:
@Thalley thanks.

What are your thoughts on stuff like `struct bt_iso_chan **cis_channels` and the array vs pointer syntax. Wouldn't the type be clearer as `struct bt_iso_chan *cis_channels[]`?

jhedberg replied:
Btw, regarding size_t vs uint8_t - it's not completely the same with ISO. There you have the additional consideration of struct packing. However, with the current layout of `bt_iso_cig_parem` there will not be any memory penalty with the changed type.

Thalley replied:
@jhedberg Added a new commit that changes the uses of uint8_t to size_t when using describing an array length.

asbjornsabo replied:
> @Thalley thanks.
> 
> What are your thoughts on stuff like `struct bt_iso_chan **cis_channels` and the array vs pointer syntax. Wouldn't the type be clearer as `struct bt_iso_chan *cis_channels[]`?

I may not be representative, but I find the latter clearer. (And ditto for [] for an array.)

Thalley committed:
Bluetooth: Audio: Use size_t for audio.h array len arguments

Use size_t instead of uint8_t to describe array lengths.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on include/bluetooth/audio/capabilities.h:
```diff
@@ -155,7 +155,7 @@ struct bt_audio_capability_ops {
 	 *  @return 0 in case of success or negative value in case of error.
 	 */
-	int (*enable)(struct bt_audio_stream *stream, uint8_t meta_count,
+	int (*enable)(struct bt_audio_stream *stream, size_t meta_count,
 		      struct bt_codec_data *meta);
```
Unrelated to this PR, but it seems this is inconsistent: in other places you have the array first and then the array size. Here it's the other way around. Any particular reason? Would be good to make this consistent throughout the codebase.

Thalley replied:
Agreed, they should be the same. I suggest to follow the convention of the standard libraries that have the array before the size/length: https://github.com/zephyrproject-rtos/zephyr/discussions/41120#discussioncomment-2242601

jhedberg commented on include/bluetooth/audio/audio.h:
```diff
@@ -1324,9 +1324,9 @@ struct bt_audio_discover_params {
 	/** Number of capabilities found */
-	uint8_t  num_caps;
+	size_t   num_caps;
 	/** Number of endpoints found */
-	uint8_t  num_eps;
+	size_t   num_eps;
```
Which arrays are these referring to? I don't seem them in the same struct?

Thalley replied:
Hmm, this one is a bit unique. 

The struct is part of the callback in 
```c
typedef void (*bt_audio_discover_func_t)(struct bt_conn *conn,
					 struct bt_codec *codec,
					 struct bt_audio_ep *ep,
					 struct bt_audio_discover_params *params);
```

where it reports the number of caps and eps it found, but where the callback is called for each endpoint there is found. So first time it will be called with `num_eps=1`, then 2, then 3, etc. 

I'll revert this one, as that need some other refactoring.

asbjornsabo commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1070,6 +1070,121 @@ int bt_audio_unicast_group_create(struct bt_audio_stream *streams,
+				       uint8_t num_stream)
+{
+	struct bt_audio_stream *tmp_stream;
+	uint8_t total_steam_cnt;
```
(Nitpicking) Missing "r" - "steam" -> "stream"

Thalley committed:
Bluetooth: Audio: Fix ordering of array size after array parameter

Make it consistent for audio.h and capabilities.h that the
array size parameter appear after the array parameter.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
This is blocking https://github.com/zephyrproject-rtos/zephyr/issues/42587 so please review :)

Casper-Bonde-Bose reviewed (approved):
LGTM
